src/module.c

```diff
@@ -2407,7 +2407,34 @@ void RM_Yield(RedisModuleCtx *ctx, int flags, const char *busy_reply) {
                 server.busy_module_yield_flags |= BUSY_MODULE_YIELD_CLIENTS;
 
             /* Let redis process events */
-            processEventsWhileBlocked();
+            if (!pthread_equal(server.main_thread_id, pthread_self())) {
+                /* If we are not in the main thread, we defer event loop processing to the main thread
+                 * after the main thread enters acquiring GIL state in order to protect the event
+                 * loop (ae.c) and avoid potential race conditions. */
+
+                int acquiring;
+                atomicGet(server.module_gil_acquring, acquiring);
+                if (!acquiring) {
+                    /* If the main thread has not yet entered the acquiring GIL state,
+                     * we attempt to wake it up and exit without waiting for it to
+                     * acquire the GIL. This avoids blocking the caller, allowing them to
+                     * continue with unfinished tasks before the next yield.
+                     * We assume the caller keeps the GIL locked. */
+                    if (write(server.module_pipe[1],"A",1) != 1) {
+                        /* Ignore the error, this is best-effort. */
+                    }
+                    goto end;
```

Comment:
```suggestion
                     yield_nesting--; return;
```

Reply:
Agree with you, `goto` can be abused, but we can't ignore its advantages because of its disadvantages.
1. simplifies error handling, easy to jump to error block.
2. jumping out of multiple nesting, although we can accomplish the same thing with other ways, but `goto` is more simple.
3. improves readability, and allows you to reuse duplicate code, as in the current example, so that if we need to exit  in another place in the future, all we need to do is `goto end`.